checks/image/image_check.go

```diff
@@ -0,0 +1,148 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package image
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/digitalocean/clusterlint/checks"
+	"github.com/digitalocean/clusterlint/kube"
+	"github.com/docker/distribution/reference"
+	corev1 "k8s.io/api/core/v1"
+)
+
+func init() {
+	checks.Register(&containerImageCheck{})
+}
+
+type containerImageCheck struct{}
+
+// Name returns a unique name for this check.
+func (pc *containerImageCheck) Name() string {
+	return "container-images"
+}
+
+// Groups returns a list of group names this check should be part of.
+func (pc *containerImageCheck) Groups() []string {
+	return []string{"image"}
+}
+
+// Description returns a detailed human-readable description of what this check
+// does.
+func (pc *containerImageCheck) Description() string {
+	return "Checks if there are pods with containers in privileged mode"
+}
+
+// Run runs this check on a set of Kubernetes objects. It can return warnings
+// (low-priority problems) and errors (high-priority problems) as well as an
+// error value indicating that the check failed to run.
+func (pc *containerImageCheck) Run(ctx context.Context, objects *kube.Objects) ([]checks.Diagnostic, error) {
+	var diagnostics []checks.Diagnostic
+	repos := ctx.Value(checks.Repositories)
+	validRepos := make(map[string]struct{})
+	if repos != nil {
+		repoArr, ok := repos.([]string)
+		if ok {
+			for _, r := range repoArr {
+				validRepos[r] = struct{}{}
+			}
+		}
+	}
+	for _, pod := range objects.Pods.Items {
+		diagnostics = append(diagnostics, pc.checkPod(validRepos, pod)...)
+	}
+
+	return diagnostics, nil
+}
+
+// checkPod checks if all pod images are well formed and
+// do not have any errors, or obvious mistakes.
+func (pc *containerImageCheck) checkPod(validRepos map[string]struct{}, pod corev1.Pod) []checks.Diagnostic {
+	var diagnostics []checks.Diagnostic
+	containers := append(pod.Spec.Containers, pod.Spec.InitContainers...)
+	for _, container := range containers {
+		named, err := checkImageError(container.Name, container.Image)
+		if err != nil {
+			diagnostics = append(diagnostics, checks.Diagnostic{
+				Check:    pc.Name(),
+				Severity: checks.Error,
+				Message:  err.Error(),
+				Kind:     checks.Pod,
+				Object:   &pod.ObjectMeta,
+				Owners:   pod.ObjectMeta.GetOwnerReferences(),
+			})
+		} else {
+			for _, e := range lintImage(named, validRepos, container.Name, container.Image) {
+				diagnostics = append(diagnostics, checks.Diagnostic{
+					Check:    pc.Name(),
+					Severity: checks.Warning,
+					Message:  e.Error(),
+					Kind:     checks.Pod,
+					Object:   &pod.ObjectMeta,
+					Owners:   pod.ObjectMeta.GetOwnerReferences(),
+				})
+			}
+		}
+	}
+	return diagnostics
+}
+
+func checkImageError(containerName, image string) (reference.Named, error) {
+	named, err := reference.ParseNormalizedNamed(image)
```

Comment:
We do a check for malformed image name in https://github.com/digitalocean/clusterlint/blob/master/checks/basic/fully_qualified_image.go. I wonder if it is necessary in both places actually. If there is an error due to malformed image or anything else, will the pod object even be available for us to check? https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.16/#podlist-v1-core just says list of pods. I'll check and update here.

Reply:
Okay, this is interesting. These should definitely get moved there. Good catch. However, if it's okay with you and @adamwg  I think we should rework the logic there a little bit. The docker `reference` package of detecting errors in image names, which should be out-and-out cluster lint errors, IMO. Thereafter non-fully-qualified images and other warnings could be added. Thoughts?